### PR TITLE
LXD nodes

### LINKED ISSUE #5: Support for LXD Nodes
See [`LXDNode.py`](../blob/9dc798cbb24346d92f450fbe6479cf8055a89144/nodes/LXDNode.py)

Original issue by @arneboockmeyer

### PR DESCRIPTION
**Pull request by mgjm**
Closes #5

### CODE REVIEW

**Comment by arneboockmeyer** [comment]
This doesn't work on my machine. I get the following error:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/urllib3/connectionpool.py", line 672, in urlopen
    chunked=chunked,
  File "/usr/local/lib/python3.7/dist-packages/urllib3/connectionpool.py", line 387, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.7/http/client.py", line 1229, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1275, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1224, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1016, in _send_output
    self.send(msg)
  File "/usr/lib/python3.7/http/client.py", line 956, in send
    self.connect()
  File "/usr/local/lib/python3.7/dist-packages/requests_unixsocket/adapters.py", line 41, in connect
    sock.connect(socket_path)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/local/lib/python3.7/dist-packages/urllib3/connectionpool.py", line 720, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/usr/local/lib/python3.7/dist-packages/urllib3/util/retry.py", line 400, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.7/dist-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.7/dist-packages/urllib3/connectionpool.py", line 672, in urlopen
    chunked=chunked,
  File "/usr/local/lib/python3.7/dist-packages/urllib3/connectionpool.py", line 387, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.7/http/client.py", line 1229, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1275, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1224, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1016, in _send_output
    self.send(msg)
  File "/usr/lib/python3.7/http/client.py", line 956, in send
    self.connect()
  File "/usr/local/lib/python3.7/dist-packages/requests_unixsocket/adapters.py", line 41, in connect
    sock.connect(socket_path)
urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/pylxd/client.py", line 308, in __init__
    response = self.api.get()
  File "/usr/local/lib/python3.7/dist-packages/pylxd/client.py", line 158, in get
    response = self.session.get(self._api_endpoint, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.7/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "examples/lxd_example.py", line 22, in <module>
    parser.run(main)
  File "/usr/local/lib/python3/dist-packages/testbed/argparse.py", line 59, in run
    main(**args)
  File "examples/lxd_example.py", line 18, in main
    sim.simulate()
  File "/usr/local/lib/python3/dist-packages/testbed/simulation.py", line 162, in simulate
    self.prepare()
  File "/usr/local/lib/python3/dist-packages/testbed/util.py", line 19, in wrapper
    result = func(*args, **kwargs)
  File "/usr/local/lib/python3/dist-packages/testbed/simulation.py", line 134, in prepare
    node.prepare(self)
  File "/usr/local/lib/python3/dist-packages/testbed/node/lxd.py", line 78, in prepare
    self.create_container()
  File "/usr/local/lib/python3/dist-packages/testbed/node/lxd.py", line 85, in create_container
    client = pylxd.Client()
  File "/usr/local/lib/python3.7/dist-packages/pylxd/client.py", line 315, in __init__
    raise exceptions.ClientConnectionFailed()
pylxd.exceptions.ClientConnectionFailed
```

Its propably some kind of configuration issue on my machine. Have you somehow modified the lxd configuration? 
Otherwise this could be an explanation: https://github.com/lxc/pylxd/issues/317

File: container-images/testbed-base/Dockerfile [hunk 1]
```diff
@@ -17,6 +17,7 @@ RUN pip3 install --no-cache-dir \
 	docker \
 	nsenter \
 	paramiko \
+	pylxd \
```
Comment (arneboockmeyer) [inline container-images/testbed-base/Dockerfile]: I would recommend to implement what I suggested in #11 before adding it here.

**Review by arneboockmeyer — COMMENTED** [review]
Full review not finished yet since its currently not testable.

**Comment by pjgoodall** [comment]
I'm just trying the most basic local test and its connection fails. This is a low-fi test. Surely it should work. 

` Ubuntu 18.04.4 LTS`

``` python
from pylxd import Client
client = Client()

all_containers = client.containers.all()

print(client.containers.all())
```


lxc installed as 
```
% sudo snap install lxd
% lxd --version
3.20
```

Python environment installed using conda from environment.yml
```
name: lxc-x11
channels:
  - anaconda
  - conda-forge
  - defaults
dependencies:
  - pip
  - pylint
  - anaconda:python=3.7
  - pip:
    - pylxd
```

```
# packages in environment at ~/miniconda3/envs/lxc-x11:
#
# Name                    Version                   Build  Channel
astroid                   2.3.3                    py37_0    anaconda
ca-certificates           2020.1.1                      0    anaconda
certifi                   2019.11.28               py37_0    anaconda
cffi                      1.14.0                   pypi_0    pypi
chardet                   3.0.4                    pypi_0    pypi
cryptography              2.8                      pypi_0    pypi
idna                      2.8                      pypi_0    pypi
isort                     4.3.21                   py37_0    anaconda
lazy-object-proxy         1.4.3            py37h7b6447c_0    anaconda
ld_impl_linux-64          2.33.1               h53a641e_7    anaconda
libedit                   3.1.20181209         hc058e9b_0    anaconda
libffi                    3.2.1                h4deb6c0_3    anaconda
libgcc-ng                 9.1.0                hdf63c60_0    anaconda
libstdcxx-ng              9.1.0                hdf63c60_0    anaconda
mccabe                    0.6.1                    py37_1    anaconda
ncurses                   6.1                  he6710b0_1    anaconda
openssl                   1.1.1                h7b6447c_0    anaconda
pbr                       5.4.4                    pypi_0    pypi
pip                       20.0.2                   py37_1    anaconda
pycparser                 2.19                     pypi_0    pypi
pylint                    2.4.4                    py37_0    anaconda
pylxd                     2.2.10                   pypi_0    pypi
python                    3.7.6                h0371630_2    anaconda
python-dateutil           2.8.1                    pypi_0    pypi
readline                  7.0                  h7b6447c_5    anaconda
requests                  2.22.0                   pypi_0    pypi
requests-toolbelt         0.9.1                    pypi_0    pypi
requests-unixsocket       0.2.0                    pypi_0    pypi
setuptools                45.1.0                   py37_0    anaconda
six                       1.14.0                   py37_0    anaconda
sqlite                    3.31.1               h7b6447c_0    anaconda
tk                        8.6.8                hbc83047_0    anaconda
urllib3                   1.25.8                   pypi_0    pypi
wheel                     0.34.2                   py37_0    anaconda
wrapt                     1.11.2           py37h7b6447c_0    anaconda
ws4py                     0.5.1                    pypi_0    pypi
xz                        5.2.4                h14c3975_4    anaconda
zlib                      1.2.11               h7b6447c_3 
```

**Comment by felix-gohla** [comment]
> ```python
> from pylxd import Client
> client = Client()
> 
> all_containers = client.containers.all()
> 
> print(client.containers.all())
> ```

Hi, 
is this related to the **testbed `cohydra`** or a general problem with Python and LXD?

Nevertheless, I tried to reproduce your issue and had no problems connecting to the LXD Daemon via `pylxd`.
Did you initialise LXD after installing it via snap?
```sh
$ lxd init
```

Without `lxd init`, there is no daemon running and you cannot connect.
If you did initialise it, could you describe your connection problems. Do you have some output from the script?

**Comment by pjgoodall** [comment]
Hi Felix,

it could be this issue: https://github.com/eclipse-fog05/fog05/issues/91

I have been using lxc for some time, and needed better control than scripting in echo zsh could give me. I was wanting to isolate development environments within lxc containers.

Containers run mostly fine.
```
 % lxc list      
+---------+---------+------+------+-----------+-----------+
|  NAME   |  STATE  | IPV4 | IPV6 |   TYPE    | SNAPSHOTS |
+---------+---------+------+------+-----------+-----------+
| byobu   | STOPPED |      |      | CONTAINER | 0         |
+---------+---------+------+------+-----------+-----------+
| gui1804 | STOPPED |      |      | CONTAINER | 1         |
+---------+---------+------+------+-----------+-----------+
```
Why am I getting connection refused. I have permissions at the command-line to perform essentially the same operation as `lxc list`

```
Traceback (most recent call last):
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1230, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1276, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1225, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1004, in _send_output
    self.send(msg)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 944, in send
    self.connect()
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/requests_unixsocket/adapters.py", line 41, in connect
    sock.connect(socket_path)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/urllib3/connectionpool.py", line 719, in urlopen
    retries = retries.increment(
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/urllib3/util/retry.py", line 400, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1230, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1276, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1225, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 1004, in _send_output
    self.send(msg)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/http/client.py", line 944, in send
    self.connect()
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/requests_unixsocket/adapters.py", line 41, in connect
    sock.connect(socket_path)
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionRefusedError(111, 'Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/pylxd/client.py", line 308, in __init__
    response = self.api.get()
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/pylxd/client.py", line 158, in get
    response = self.session.get(self._api_endpoint, *args, **kwargs)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionRefusedError(111, 'Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./bin/pylxd-test.py", line 2, in <module>
    client = Client()
  File "/home/peterg/miniconda3/envs/lxc-x11/lib/python3.8/site-packages/pylxd/client.py", line 315, in __init__
    raise exceptions.ClientConnectionFailed()
pylxd.exceptions.ClientConnectionFailed

```

> > ```python
> > from pylxd import Client
> > client = Client()
> > 
> > all_containers = client.containers.all()
> > 
> > print(client.containers.all())
> > ```
> 
> Hi,
> is this related to the **testbed `cohydra`** or a general problem with Python and LXD?
> 
> Nevertheless, I tried to reproduce your issue and had no problems connecting to the LXD Daemon via `pylxd`.
> Did you initialise LXD after installing it via snap?
> 
> ```shell
> $ lxd init
> ```
> 
> Without `lxd init`, there is no daemon running and you cannot connect.
> If you did initialise it, could you describe your connection problems. Do you have some output from the script?

**Comment by felix-gohla** [comment]
Hi Peter,

indeed, this is a weird issue you're experiencing. I can't reproduce it.

Did you try to follow the way proposed [here](https://github.com/lxc/pylxd/issues/365#issuecomment-500465255)?

Nevertheless again the question: How is this related to `cohydra`? Cohydra is a testbed for network simulation - just a project *using* `pylxd`, so maybe you try to open an issue at the [*official* pylxd repository](https://github.com/lxc/pylxd/)?

**Comment by pjgoodall** [comment]
Felix,

Oh I’m sorry, i had better move my report to the correct forum. Sorry for
the distraction. Thanks 😊

**Comment by arneboockmeyer** [comment]
@rs22 I think the build failed because we haven't updatet the docker image on docker hub, right? Do you know, how to update that one?

**Comment by rs22** [comment]
The ('latest', 'dev' and 'base') docker images are automatically updated on every commit to `master`. Also, there is a separate tag available for each commit. As you can see, we have 'base' images (required for docs build) for the current master commit: https://hub.docker.com/repository/docker/osmhpi/cohydra/tags?page=1

The problem is that the pylxd dependency introduced in this PR is not yet contained in the Docker image (which is always built from master). See also the diff of `docker/cohydra-base/Dockerfile`.

It's a bit cumbersome that we have two different CI build systems and configurations that are not synchronized (i.e. the base Docker image is not build locally before the docs build). You could work around this for now by creating and merging a separate PR that only introduces the additional pylxd dependency in the Dockerfile and wait for the new 'base'-image to be built by Docker Hub.

**Comment by arneboockmeyer** [comment]
@rs22 thanks for the explanation. Look at this one: https://github.com/osmhpi/cohydra/pull/40

**Comment by arneboockmeyer** [comment]
@rs22 @felix-gohla The build works now. @rs22 please review

**Review by rs22 — APPROVED** [review]

**Pull request merged**